### DISCUSSION TITLE
Setting up mailcrypt when using OAuth2 for authentication

### DISCUSSION THREAD

**Discussion by ghost** [docker-mailserver/docker-mailserver#3888] [Q&A]
I've been using docker-mailserver for a while and already have single sign-on set up with Keycloak using OAuth2. I've completely disabled plaintext authentication because I was a little worried about the security (I've already seen a few bots try to brute force over SMTP), but now I've been thinking it would also be a good idea to try using mailcrypt to encrypt my emails on disk.

I know that you need the user password in order to decrypt emails stored on disk, and I also know that (at least in the official recommended steps to take on dovecot.org) that you're supposed to have a username/password database set up for mailcrypt to use. I don't want to break something in my mailserver from just messing around with this—has anybody tried using mailcrypt and OAuth2 at the same time? What should I do?

I've already looked through [this](https://github.com/docker-mailserver/docker-mailserver/issues/1171#issuecomment-805969981) thread, which is where I got some clarification on the database requirement, but it still seems like after all the work they did in that thread that a password is required.

The optimal flow in my eyes is that I can authenticate using OAuth2, but then I'm given just a password box so that I can type in a password to decrypt my mail. If anyone here is experienced with Vaultwarden, I think it should work like that, where OAuth2 doesn't get you in to the encrypted files, it just works in place of the "username/email" box.

For the client that I usually use to connect to the mailserver, I use Roundcube.

Just as a note, I don't need any emails that I already have to be encrypted or anything like that—I don't have any real major security or privacy concerns that mean that I need to do this, either—I just think that if I can, then I should, and I like learning about these things.

**Comment by polarathene** [comment]
> I've already seen a few bots try to brute force over SMTP

If this is a personal mail server, you shouldn't have to worry if you use a strong password (high entropy).

A password manager is good for that, but something like [`detailed snail summons slim lab coat` can be quite easy to input and safe](https://github.com/docker-mailserver/docker-mailserver/issues/3863#issuecomment-1922733241) (_when generated properly_), the cost to attack it would be too expensive unless the breach was targeting you specifically for something more valuable than the cost.

---

I still have the mail encryption feature support on my TODO, it's scheduled after I wrap up the LDAP rework (_which has been stuck on finding time to rewrite related docs_). Hopefully this month I'll finally get that in 😓 

The feature can be either:
- Global encryption key. No password needed, but an admin or server compromise could decrypt the files.
- Per-user encryption keys. When the user logs in their password is the decryption key, preventing anyone with server access from decrypting your mail.

With OAuth2, you login with your username and an access token. The access token is provided from your login session with the auth service handling your identity, then Dovecot will check with that auth service, giving it the access token received to confirm the token is still valid and that the email address associated to it is your login username (_unless Dovecot was configured differently by the DMS admin_).

The access token is dynamic, so it can't be used to decrypt the mail. We don't control mail clients, AFAIK there is no way to prompt for a password from DMS/Dovecot, the information is expected at login implicitly.

I am pretty sure that you can provide the password via other means though. So technically the auth service should be capable of returning other data beyond the email address. If it can include a user encryption key that could be used as input instead for decrypting, Dovecot just needs additional configuration to know about the field and add it into the PassDB for OAuth2.

The actual key for decrypting your mail is protected via another layer of encryption IIRC. This is so that mail is only ever encrypted once to disk, and that if you change your password, we only need to know the original password (hashed) and the new one (hashed) to "change the lock" for the decryption key.

You'll have a problem there with the OAuth2 process if that value is different from your hashed password. Due to that process above, you can support multiple secrets to encrypt the decryption key, but we only provide a CLI for password change to update it. With OAuth2, we could only query the active secret for that user, so if it differs from the password hash, and the auth service changed it, we don't have the original information to update/swap for the new secret...I'm terrible at explaining things 😆 

Anyway, good news is it should be possible with some extra config. It just may not work well if the secret was to ever change (_technically if it's not associated to a users password, and has enough entropy it shouldn't really need to change unless their was a compromise/leak_).

Just keep in mind that if the input required to decrypt the mail is lost with per-user keys, it's effectively impossible to recover from as the encryption strength would require enough energy to boil all the oceans in the world 🤯

**Reply by ghost** [reply]
Hello, thanks for the response! I'll try to keep this short, because I don't have much else to add to the conversation and it sounds like the use-case that I was talking about is already on your mind.

The exact thing that I was worried about was that (because I've already set up OAuth2) the lack of a password input would mean that I wouldn't have a chance at providing a password during login to decrypt the mail. I didn't know whether the password prompt was something that the mail client generated on request from the mail server or if the client just does all the information-gathering and sends the data all at once to the mail server (it sounds to me like it's the second one, thanks for the clarification!). You aren't terrible at explaining things, I actually think I'm picking up the core concepts pretty quickly just reading what you wrote. Thanks again!

Since I'm just a hobbyist and not super-knowledgable beyond an overview level, and I don't particularly want to run the risk of driving a wedge into my personal mail server, I think I won't mess around too much and I might just wait until better documentation / an official feature is available.

Right now, it seems like my best option would be to use a server-wide key so that no additional information is needed during login. Since this is in Docker, I could just "hide" the encryption key inside the container—that way, the key would just vanish if the system were shut down, container stopped, only the folder with mounted volumes was stolen, etc. I could just wipe all the old emails away and start fresh—if the information's valuable, I usually store it somewhere more secure than my email inbox anyway.

I appreciate your response, again, and thanks for such a cool project!
